src/ray/core_worker/transport/direct_actor_transport.cc

```diff
@@ -47,85 +44,50 @@ Status CoreWorkerDirectActorTaskSubmitter::SubmitTask(
     // specified time.
     pending_requests_[actor_id].emplace_back(std::move(request));
     RAY_LOG(DEBUG) << "Actor " << actor_id << " is not yet created.";
-  } else if (iter->second.state_ == ActorTableData::ALIVE) {
-    // Actor is alive, submit the request.
-    if (rpc_clients_.count(actor_id) == 0) {
-      // If rpc client is not available, then create it.
-      ConnectAndSendPendingTasks(actor_id, iter->second.location_.first,
-                                 iter->second.location_.second);
-    }
-
-    // Submit request.
-    auto &client = rpc_clients_[actor_id];
-    PushTask(*client, std::move(request), actor_id, task_id, num_returns);
   } else {
-    // Actor is dead, treat the task as failure.
-    RAY_CHECK(iter->second.state_ == ActorTableData::DEAD);
-    TreatTaskAsFailed(task_id, num_returns, rpc::ErrorType::ACTOR_DIED);
+    // Submit request.
+    PushTask(*rpc_client, std::move(request), actor_id, task_id, num_returns);
   }
 
   // If the task submission subsequently fails, then the client will receive
   // the error in a callback.
   return Status::OK();
 }
 
-void CoreWorkerDirectActorTaskSubmitter::HandleActorUpdate(
-    const ActorID &actor_id, const ActorTableData &actor_data) {
+void CoreWorkerDirectActorTaskSubmitter::FailPendingTasks(const ActorID &actor_id) {
   std::unique_lock<std::mutex> guard(mutex_);
-  actor_states_.erase(actor_id);
-  actor_states_.emplace(
-      actor_id,
-      ActorStateData(actor_data.state(), actor_data.ip_address(), actor_data.port()));
-
-  if (actor_data.state() == ActorTableData::ALIVE) {
-    // Check if this actor is the one that we're interested, if we already have
-    // a connection to the actor, or have pending requests for it, we should
-    // create a new connection.
-    if (pending_requests_.count(actor_id) > 0 && rpc_clients_.count(actor_id) == 0) {
-      ConnectAndSendPendingTasks(actor_id, actor_data.ip_address(), actor_data.port());
-    }
-  } else {
-    // Remove rpc client if it's dead or being reconstructed.
-    rpc_clients_.erase(actor_id);
-
-    // For tasks that have been sent and are waiting for replies, treat them
-    // as failed when the destination actor is dead or reconstructing.
-    auto iter = waiting_reply_tasks_.find(actor_id);
-    if (iter != waiting_reply_tasks_.end()) {
-      for (const auto &entry : iter->second) {
-        const auto &task_id = entry.first;
-        const auto num_returns = entry.second;
-        TreatTaskAsFailed(task_id, num_returns, rpc::ErrorType::ACTOR_DIED);
-      }
-      waiting_reply_tasks_.erase(actor_id);
+  // For tasks that have been sent and are waiting for replies, treat them
```

Comment:
So it's not pending tasks more like fail in flight tasks right?

Actually do we even need this? The RPC requests will fail by themselves right?

Reply:
I guess the old code did this already, maybe it was just to fail them faster. @zhijunfu ?

Reply:
@ericl I remember the issue was that it's not guaranteed that GRPC callback will be invoked on failure case, so we also leverage the GCS actor notification to treat tasks as failed.